testing/unittests/jthreadpooltests.cpp

```diff
@@ -0,0 +1,943 @@
+        duration = measureThreadStartDuration(startTime, 2);
+        CPPUNIT_ASSERT(duration.count() < 100);
+
+        // Test: Pool was above, fell below, and is now at defaultmax again
+        CPPUNIT_ASSERT_EQUAL(3U, pool->runningCount());
+
+        // All threads should complete
+        pool->joinAll(true);
+        CPPUNIT_ASSERT_EQUAL(7U, threadStartedCount.load());
+        CPPUNIT_ASSERT_EQUAL(7U, threadCompletedCount.load());
+        CPPUNIT_ASSERT_EQUAL(0U, pool->runningCount());
+        CPPUNIT_ASSERT(pool->running());
+    }
+
+    void testThrottledThreadPoolWithFastThreadCompletion()
+    {
+        // Test the behavior of a throttled thread pool with 100 default max threads
+        // and 1ms default start delay under high load with fast-completing threads
+        pool.setown(createThreadPool("TestPoolWithFastThreadCompletion", factory, true, nullptr, 100, 1));
+
+        for (unsigned numIterations = 1; numIterations <= 10; numIterations++)
+        {
+            threadStartedCount = 0;
+            threadCompletedCount = 0;
+
+            unsigned threadId{0};
+            StringBuffer threadName;
+            auto startTime = std::chrono::high_resolution_clock::now();
+            for (unsigned threadId = 1; threadId <= 200; threadId++)
+            {
+                threadName.appendf("Thread%u", threadId);
+                if (threadId < 101)
+                    params = {lifespan500milliseconds, threadName.str()};
+                else
+                    params = {lifespan1000milliseconds, threadName.str()};
+                pool->start(&params, params.name);
+            }
+
+            auto duration = measureThreadStartDuration(startTime, 200);
+            CPPUNIT_ASSERT(duration.count() < 500);
+            CPPUNIT_ASSERT_EQUAL(200U, pool->runningCount());
+            CPPUNIT_ASSERT_EQUAL(200U, threadStartedCount.load());
+            CPPUNIT_ASSERT_EQUAL(0U, threadCompletedCount.load());
+
+            // Wait for 100 threads to complete, leaving the pool at capacity
+            while (pool->runningCount() > 100)
+                std::this_thread::sleep_for(std::chrono::milliseconds(10));
+
+            // Test: Pool is at capacity, should throttle new thread starts
+            //
+            unsigned exceptionCount{0};
+            // 100 threads are still running that have approx. 500 milliseconds of runtime left
+            // so with 100ms thread pool default start delay we can expect 5 threads
+            // to start with no delay. Whilst 100 threads will wait up to 100ms before
+            // timing out.
+            startTime = std::chrono::high_resolution_clock::now();
+            for (threadId = 201; threadId <= 400; threadId++)
+            {
+                threadName.appendf("Thread%u", threadId);
+                params = {lifespan1000milliseconds, threadName.str()};
+                attemptThreadStartWithTimeout(&params, 100, "TestPoolWithFastThreadCompletion", exceptionCount,
```

Comment:
re. earlier comment about attemptThreadStartWithTimeout returning success or timeouit..
I do think it would be cleaner if exceptionCount was not a passed in incremented reference but a return state, e.g.:
```
if (!attemptThreadStartWithTimeout(....)
    exceptionCount++;
```